common/flatpak-run.c

```diff
@@ -3791,9 +3833,48 @@ flatpak_run_app (FlatpakDecomposed *app_ref,
 
   flatpak_bwrap_add_args (bwrap,
                           "--ro-bind", flatpak_file_get_path_cached (runtime_files), "/usr",
-                          "--lock-file", "/usr/.ref",
                           NULL);
 
+  if (runtime_files == original_runtime_files)
+    {
+      /* All true Flatpak runtimes have files/.ref */
+      flatpak_bwrap_add_args (bwrap,
+                              "--lock-file", "/usr/.ref",
+                              NULL);
+    }
+  else
+    {
+      g_autoptr(GFile) runtime_child = NULL;
+
+      runtime_child = g_file_get_child (runtime_files, ".ref");
+
+      /* Lock ${usr}/.ref if it exists */
+      if (g_file_query_exists (runtime_child, NULL))
+        flatpak_bwrap_add_args (bwrap,
+                                "--lock-file", "/usr/.ref",
+                                NULL);
+
+      /* Put the real Flatpak runtime in /run/runtime, so that the
+       * replacement /usr can have symlinks into /run/runtime in order
+       * to use the Flatpak runtime's graphics drivers etc. if desired */
+      flatpak_bwrap_add_args (bwrap,
+                              "--ro-bind",
+                              flatpak_file_get_path_cached (original_runtime_files),
+                              "/run/runtime/usr",
+                              "--lock-file", "/run/runtime/usr/.ref",
+                              NULL);
+      flatpak_run_setup_usr_links (bwrap, original_runtime_files,
+                                   "/run/runtime");
+
+      g_clear_object (&runtime_child);
+      runtime_child = g_file_get_child (runtime_files, "etc");
+
+      if (g_file_query_exists (runtime_child, NULL))
+        flatpak_bwrap_add_args (bwrap,
+                                "--symlink", "usr/etc", "/run/runtime/etc",
+                                NULL);
```

Comment:
This will be a bit different from the regular /etc that would be if this was used as a runtime in the regular way. Still, probably right to have it there if there are things like relative symlinks pointing to it.

Reply:
Right; in particular, if we want to look at the runtime's `/etc/ld.so.cache`, we can't, unless we provide `/etc`. It would be possible to do the full `/etc` setup for both `/etc` and `/run/runtime/etc`, but that seemed like extra code for no real gain.

Reply:
(This is unchanged except for moving it to `/run/parent/etc`; I think it's still correct?)